### PR TITLE
fix: allow accessing file:// when web security is disabled

### PR DESCRIPTION
**Pull request by zcbenz**
#### Description of Change

Subresource `file://` requests should be explicitly allowed when web security is disabled, closes https://github.com/electron/electron/issues/23757.

This PR also rewrites webview web security tests, since they were passing even though `file://` requests were actually failing.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fix failing to request file:// resources when web security is disabled.

### CODE REVIEW

**Review by codebytere — APPROVED** [review]

File: shell/browser/protocol_registry.cc [hunk 1]
```diff
@@ -72,11 +75,18 @@ void ProtocolRegistry::RegisterURLLoaderFactories(
     DCHECK(!base::Contains(*factories, url::kFileScheme));
     factories->emplace(url::kFileScheme, AsarURLLoaderFactory::Create());
   } else if (type == URLLoaderFactoryType::kDocumentSubResource) {
-    // Only support requesting file:// subresource URLs when Chromium does so,
-    // it is usually supported under file:// or about:blank documents.
+    // Only support requesting file:// subresource URLs when:
+    // 1. it is supported by Chromium browser, such as under file:// or
+    //    about:blank documents;
+    // 2. web security is turned off.
```
Comment (nornagon) [inline shell/browser/protocol_registry.cc:81]: this is (1) OR (2), right?
```suggestion
    // 1. it is supported by Chromium browser, such as under file:// or
    //    about:blank documents; or
    // 2. web security is turned off.
```

File: shell/browser/protocol_registry.h [hunk 2]
```diff
@@ -27,7 +27,8 @@ class ProtocolRegistry {
 
   void RegisterURLLoaderFactories(
       URLLoaderFactoryType type,
-      content::ContentBrowserClient::NonNetworkURLLoaderFactoryMap* factories);
+      content::ContentBrowserClient::NonNetworkURLLoaderFactoryMap* factories,
+      content::WebContents* web_contents = nullptr);
```
Comment (nornagon) [inline shell/browser/protocol_registry.h:31]: nit: I think we can [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) this a bit and just pass a bool for whether to allow file scheme access:

```suggestion
      bool allow_file_access = false);
```

& then move the WebPreferences logic into ElectronBrowserClient::RegisterNonNetworkSubresourceURLLoaderFactories

File: spec-main/chromium-spec.ts [hunk 3]
```diff
@@ -287,6 +287,39 @@ describe('web security', () => {
...
+      return await w.webContents.executeJavaScript('loadFile()');
+    }
+
+    it('is forbidden when web security is enabled', async () => {
+      const w = new BrowserWindow({ show: false, webPreferences: { webSecurity: true } });
+      const result = await loadFile(w);
+      expect(result).to.equal('failed');
+    });
+
+    it('is forbidden when web security is disabled', async () => {
```
Comment (nornagon) [inline spec-main/chromium-spec.ts:316]:
```suggestion
    it('is allowed when web security is disabled', async () => {
```

File: spec/webview-spec.js [hunk 4]
```diff
@@ -321,27 +343,13 @@ describe('<webview> tag', function () {
 
   describe('disablewebsecurity attribute', () => {
     it('does not disable web security when not set', async () => {
-      const jqueryPath = path.join(__dirname, '/static/jquery-2.0.3.min.js');
```
Comment (nornagon) [inline spec/webview-spec.js:324]: Can we delete jquery-2.0.3.min.js from our tests now?

Comment (zcbenz) [reply spec/webview-spec.js:324]: There are still a few places using it, but I think we should be able to rework those places to remove this file.

**Review by nornagon — APPROVED** [review]

**Comment by jtbandes** [comment]
It looks like this PR also fixed https://github.com/electron/electron/issues/28572! 🙌  Was that an intentional side-effect of this change? Would it make sense to add more tests to ensure the SharedWorker behavior specifically does not regress?

File: shell/browser/electron_browser_client.cc [hunk 5]
```diff
@@ -1297,8 +1298,10 @@ void ElectronBrowserClient::
         NonNetworkURLLoaderFactoryMap* factories) {
   auto* protocol_registry =
       ProtocolRegistry::FromBrowserContext(browser_context);
-  protocol_registry->RegisterURLLoaderFactories(
-      URLLoaderFactoryType::kWorkerMainResource, factories);
+  // Workers are not allowed to request file:// URLs, there is no particular
+  // reason for it, and we could consider supporting it in future.
```
Comment (jtbandes) [inline shell/browser/electron_browser_client.cc:1302]: This comment makes me wonder if the fix for SharedWorkers was unintentional 😛

**Pull request merged**